File: server/scanner.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { parse as parseToml } from 'toml';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// code_workshop 디렉토리 경로 (이 프로젝트의 부모 디렉토리)
const WORKSPACE_PATH = path.join(__dirname, '../..');

// 스캔에서 제외할 디렉토리
const EXCLUDED_DIRS = [
  '.', '..', '.DS_Store', '__pycache__', 'node_modules',
  '.git', '.venv', 'venv', '.claude', '_project-dashboard'
];

/**
 * README.md에서 설명 추출
 */
function extractDescription(readmePath) {
  try {
    const content = fs.readFileSync(readmePath, 'utf-8');
    const lines = content.split('\n').filter(line => line.trim() !== '');

    // 첫 번째 제목(#) 다음의 텍스트 추출
    let description = '';
    let foundTitle = false;

    for (const line of lines) {
      if (line.startsWith('#')) {
        foundTitle = true;
        // 제목에서 # 제거하고 설명으로 사용
        description = line.replace(/^#+\s*/, '').trim();
        continue;
      }

      if (foundTitle && !line.startsWith('#') && line.length > 10) {
        description += ' ' + line.trim();
        break;
      }
    }

    return description.slice(0, 150) || '설명 없음';
  } catch (error) {
    return '설명 없음';
  }
}

/**
 * Git remote URL 추출
 */
function extractGitRemote(projectPath) {
  try {
    const gitConfigPath = path.join(projectPath, '.git', 'config');
    if (!fs.existsSync(gitConfigPath)) return null;

    const config = fs.readFileSync(gitConfigPath, 'utf-8');
    const match = config.match(/url\s*=\s*(.+)/);
    return match ? match[1].trim() : null;
  } catch (error) {
    return null;
  }
}

/**
 * 마지막 수정 날짜 가져오기
 */
function getLastModifiedDate(projectPath) {
  try {
    // .git/logs/HEAD가 있으면 마지막 커밋 날짜 사용
    const gitLogPath = path.join(projectPath, '.git', 'logs', 'HEAD');
    if (fs.existsSync(gitLogPath)) {
      const logs = fs.readFileSync(gitLogPath, 'utf-8').trim().split('\n');
      const lastLog = logs[logs.length - 1];
      const match = lastLog.match(/>\s+(\d+)/);
      if (match) {
        return new Date(parseInt(match[1]) * 1000).toISOString();
      }
    }

    // 그렇지 않으면 디렉토리 수정 시간 사용
    const stats = fs.statSync(projectPath);
    return stats.mtime.toISOString();
  } catch (error) {
    return new Date().toISOString();
  }
}

/**
 * 단일 프로젝트 스캔
 */
function scanProject(projectPath) {
  const projectName = path.basename(projectPath);

  const project = {
    name: projectName,
    path: projectPath,
    type: [],
    description: '설명 없음',
    techStack: [],
    lastModified: getLastModifiedDate(projectPath),
    gitRemote: null,
    hasTests: false,
    hasCI: false
  };

  // package.json 확인 (Node.js/TypeScript)
  const packageJsonPath = path.join(projectPath, 'package.json');
  if (fs.existsSync(packageJsonPath)) {
    try {
      const pkg = JSON.parse(fs.readFileSync(packageJsonPath, 'utf-8'));
      project.type.push('nodejs');

      if (pkg.description) {
        project.description = pkg.description;
      }

      // 기술 스택 추출
      const deps = { ...pkg.dependencies, ...pkg.devDependencies };
      project.techStack = Object.keys(deps).slice(0, 10);

      // 테스트 확인
      if (pkg.scripts && (pkg.scripts.test || pkg.scripts['test:unit'])) {
        project.hasTests = true;
      }
    } catch (error) {
      console.error(`Error parsing package.json in ${projectName}:`, error.message);
    }
  }

  // pyproject.toml 확인 (Python Poetry)
  const pyprojectPath = path.join(projectPath, 'pyproject.toml');
  if (fs.existsSync(pyprojectPath)) {
    try {
      const content = fs.readFileSync(pyprojectPath, 'utf-8');
      const config = parseToml(content);

      project.type.push('python-poetry');

      if (config.tool?.poetry?.description) {
        project.description = config.tool.poetry.description;
      }

      // 의존성 추출
      if (config.tool?.poetry?.dependencies) {
        const deps = Object.keys(config.tool.poetry.dependencies).filter(d => d !== 'python');
        project.techStack = [...project.techStack, ...deps.slice(0, 10)];
      }
    } catch (error) {
      console.error(`Error parsing pyproject.toml in ${projectName}:`, error.message);
    }
  }

  // requirements.txt 확인 (Python pip)
  const requirementsPath = path.join(projectPath, 'requirements.txt');
  if (fs.existsSync(requirementsPath)) {
    try {
      const content = fs.readFileSync(requirementsPath, 'utf-8');
      const deps = content.split('\n')
        .map(line => line.trim())
        .filter(line => line && !line.startsWith('#'))
        .map(line => line.split(/[=<>]/)[0])
        .slice(0, 10);

      project.type.push('python-pip');
      project.techStack = [...project.techStack, ...deps];
    } catch (error) {
      console.error(`Error parsing requirements.txt in ${projectName}:`, error.message);
    }
  }

  // README.md 확인
  const readmePath = path.join(projectPath, 'README.md');
  if (fs.existsSync(readmePath) && project.description === '설명 없음') {
    project.description = extractDescription(readmePath);
  }

  // Git 정보
  project.gitRemote = extractGitRemote(projectPath);

  // CI 파일 확인
  const ciPaths = [
    path.join(projectPath, '.github', 'workflows'),
    path.join(projectPath, '.gitlab-ci.yml'),
    path.join(projectPath, '.travis.yml')
  ];

  project.hasCI = ciPaths.some(p => fs.existsSync(p));

  // 테스트 폴더 확인
  const testDirs = ['test', 'tests', '__tests__', 'spec'];
  if (!project.hasTests) {
    project.hasTests = testDirs.some(dir =>
      fs.existsSync(path.join(projectPath, dir))
    );
  }

  return project;
}

/**
 * 모든 프로젝트 스캔
 */
export function scanAllProjects() {
  console.log(`Scanning projects in: ${WORKSPACE_PATH}`);

  const projects = [];
  const entries = fs.readdirSync(WORKSPACE_PATH);

  for (const entry of entries) {
    if (EXCLUDED_DIRS.includes(entry)) continue;

    const fullPath = path.join(WORKSPACE_PATH, entry);

    try {
      const stats = fs.statSync(fullPath);
      if (stats.isDirectory()) {
        const project = scanProject(fullPath);
        projects.push(project);
      }
    } catch (error) {
      console.error(`Error scanning ${entry}:`, error.message);
    }
  }

  console.log(`Found ${projects.length} projects`);
  return projects;
}

File: src-tauri/src/scanner.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::{Path, PathBuf};
use toml::Value;

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Project {
    pub name: String,
    pub path: String,
    #[serde(rename = "type")]
    pub project_type: Vec<String>,
    pub description: String,
    pub tech_stack: Vec<String>,
    pub last_modified: String,
    pub git_remote: Option<String>,
    pub has_tests: bool,
    pub has_ci: bool,
}

const EXCLUDED_DIRS: &[&str] = &[
    ".",
    "..",
    ".DS_Store",
    "__pycache__",
    "node_modules",
    ".git",
    ".venv",
    "venv",
    ".claude",
    "_project-dashboard",
    "code-tare-native",
];

/// README.md에서 설명 추출
fn extract_description(readme_path: &Path) -> String {
    if let Ok(content) = fs::read_to_string(readme_path) {
        let lines: Vec<&str> = content.lines().filter(|l| !l.trim().is_empty()).collect();
        let mut description = String::new();
        let mut found_title = false;

        for line in lines {
            if line.starts_with('#') {
                found_title = true;
                let title = line.trim_start_matches('#').trim();
                description = title.to_string();
                continue;
            }

            if found_title && !line.starts_with('#') && line.len() > 10 {
                description.push(' ');
                description.push_str(line.trim());
                break;
            }
        }

        if description.is_empty() {
            return "설명 없음".to_string();
        }

        description.chars().take(150).collect()
    } else {
        "설명 없음".to_string()
    }
}

/// Git remote URL 추출
fn extract_git_remote(project_path: &Path) -> Option<String> {
    let git_config_path = project_path.join(".git").join("config");

    if let Ok(content) = fs::read_to_string(git_config_path) {
        for line in content.lines() {
            if line.trim().starts_with("url") {
                if let Some(url) = line.split('=').nth(1) {
                    return Some(url.trim().to_string());
                }
            }
        }
    }

    None
}

/// 마지막 수정 날짜 가져오기
fn get_last_modified_date(project_path: &Path) -> String {
    // .git/logs/HEAD에서 마지막 커밋 날짜 추출
    let git_log_path = project_path.join(".git").join("logs").join("HEAD");

    if let Ok(content) = fs::read_to_string(git_log_path) {
        if let Some(last_line) = content.lines().last() {
            // 로그 형식: ... > timestamp ...
            if let Some(timestamp_str) = last_line.split('>').nth(1) {
                if let Some(timestamp) = timestamp_str.trim().split_whitespace().next() {
                    if let Ok(ts) = timestamp.parse::<i64>() {
                        use chrono::{DateTime, Utc};
                        let dt = DateTime::<Utc>::from_timestamp(ts, 0);
                        if let Some(datetime) = dt {
                            return datetime.to_rfc3339();
                        }
                    }
                }
            }
        }
    }

    // 디렉토리 수정 시간 사용
    if let Ok(metadata) = fs::metadata(project_path) {
        if let Ok(modified) = metadata.modified() {
            use chrono::{DateTime, Utc};
            let datetime: DateTime<Utc> = modified.into();
            return datetime.to_rfc3339();
        }
    }

    use chrono::Utc;
    Utc::now().to_rfc3339()
}

/// 단일 프로젝트 스캔
fn scan_project(project_path: &Path) -> Project {
    let project_name = project_path
        .file_name()
        .unwrap_or_default()
        .to_string_lossy()
        .to_string();

    let mut project = Project {
        name: project_name.clone(),
        path: project_path.to_string_lossy().to_string(),
        project_type: Vec::new(),
        description: "설명 없음".to_string(),
        tech_stack: Vec::new(),
        last_modified: get_last_modified_date(project_path),
        git_remote: None,
        has_tests: false,
        has_ci: false,
    };

    // package.json 확인 (Node.js)
    let package_json_path = project_path.join("package.json");
    if package_json_path.exists() {
        if let Ok(content) = fs::read_to_string(&package_json_path) {
            if let Ok(pkg) = serde_json::from_str::<serde_json::Value>(&content) {
                project.project_type.push("nodejs".to_string());

                if let Some(desc) = pkg.get("description").and_then(|v| v.as_str()) {
                    project.description = desc.to_string();
                }

                // 기술 스택 추출
                let mut deps = Vec::new();
                if let Some(dependencies) = pkg.get("dependencies").and_then(|v| v.as_object()) {
                    deps.extend(dependencies.keys().cloned());
                }
                if let Some(dev_deps) = pkg.get("devDependencies").and_then(|v| v.as_object()) {
                    deps.extend(dev_deps.keys().cloned());
                }
                project.tech_stack = deps.iter().take(10).map(|s| s.to_string()).collect();

                // 테스트 확인
                if let Some(scripts) = pkg.get("scripts").and_then(|v| v.as_object()) {
                    if scripts.contains_key("test") || scripts.contains_key("test:unit") {
                        project.has_tests = true;
                    }
                }
            }
        }
    }

    // pyproject.toml 확인 (Python Poetry)
    let pyproject_path = project_path.join("pyproject.toml");
    if pyproject_path.exists() {
        if let Ok(content) = fs::read_to_string(&pyproject_path) {
            if let Ok(config) = content.parse::<Value>() {
                project.project_type.push("python-poetry".to_string());

                if let Some(desc) = config
                    .get("tool")
                    .and_then(|t| t.get("poetry"))
                    .and_then(|p| p.get("description"))
                    .and_then(|d| d.as_str())
                {
                    project.description = desc.to_string();
                }

                // 의존성 추출
                if let Some(deps) = config
                    .get("tool")
                    .and_then(|t| t.get("poetry"))
                    .and_then(|p| p.get("dependencies"))
                    .and_then(|d| d.as_table())
                {
                    let dep_list: Vec<String> = deps
                        .keys()
                        .filter(|k| k.as_str() != "python")
                        .take(10)
                        .map(|k| k.to_string())
                        .collect();
                    project.tech_stack.extend(dep_list);
                }
            }
        }
    }

    // requirements.txt 확인 (Python pip)
    let requirements_path = project_path.join("requirements.txt");
    if requirements_path.exists() {
        if let Ok(content) = fs::read_to_string(&requirements_path) {
            let deps: Vec<String> = content
                .lines()
                .map(|l| l.trim())
                .filter(|l| !l.is_empty() && !l.starts_with('#'))
                .filter_map(|l| {
                    l.split(&['=', '<', '>'][..])
                        .next()
                        .map(|s| s.to_string())
                })
                .take(10)
                .collect();

            project.project_type.push("python-pip".to_string());
            project.tech_stack.extend(deps);
        }
    }

    // README.md 확인
    let readme_path = project_path.join("README.md");
    if readme_path.exists() && project.description == "설명 없음" {
        project.description = extract_description(&readme_path);
    }

    // Git 정보
    project.git_remote = extract_git_remote(project_path);

    // CI 파일 확인
    let ci_paths = vec![
        project_path.join(".github").join("workflows"),
        project_path.join(".gitlab-ci.yml"),
        project_path.join(".travis.yml"),
    ];
    project.has_ci = ci_paths.iter().any(|p| p.exists());

    // 테스트 폴더 확인
    if !project.has_tests {
        let test_dirs = vec!["test", "tests", "__tests__", "spec"];
        project.has_tests = test_dirs
            .iter()
            .any(|dir| project_path.join(dir).exists());
    }

    project
}

/// 모든 프로젝트 스캔
pub fn scan_all_projects(workspace_path: &str) -> Result<Vec<Project>, String> {
    println!("Scanning projects in: {}", workspace_path);

    let workspace = Path::new(workspace_path);
    if !workspace.exists() {
        return Err(format!("Workspace path does not exist: {}", workspace_path));
    }

    let mut projects = Vec::new();

    let entries = fs::read_dir(workspace)
        .map_err(|e| format!("Failed to read workspace directory: {}", e))?;

    for entry in entries {
        if let Ok(entry) = entry {
            let file_name = entry.file_name().to_string_lossy().to_string();

            if EXCLUDED_DIRS.contains(&file_name.as_str()) {
                continue;
            }

            let path = entry.path();
            if path.is_dir() {
                let project = scan_project(&path);
                projects.push(project);
            }
        }
    }

    println!("Found {} projects", projects.len());
    Ok(projects)
}
